### PR TITLE
Optimizations and Refractoring

### PR DESCRIPTION
**Pull request by green-real**
Added second pass when reading instructions, after constants have been defined, to replace constant table indices in instructions with direct references to the constant. Constant is held in inst.K, besides for imports, where the constants are in inst.K0, inst.K1, inst.K2 depending on the import path size.
Updated VM execute to make use of this, replacing most uses of the constants table while executing instructions.

Renamed variables and functions to use camelCase. Some variables stay the same to match the names Luau lvmload.cpp uses.

Localized buffer functions that are being used.

### CODE REVIEW

File: Source.lua [hunk 1]
```diff
@@ -1,229 +1,281 @@
...
-		local int = buffer.readu32(stream, position)
+	local function readWord()
+		local word = buffer_readu32(stream, position)
 		position = position + 4
-		return int
+		return word
 	end
 
-	local function read_varint()
+	local function readVarint()
```
Comment (TheGreatSageEqualToHeaven) [inline Source.lua:144]: Should be `readVarInt` or `readVariableInteger`

File: Source.lua [hunk 2]
```diff
@@ -1,229 +1,281 @@
...
 			position = position + size
+
+			return str
 		end
-		return str
 	end
 
-	local luauVersion = read_byte()
+	local luauVersion = readByte()
 	local typesversion = 0
```
Comment (TheGreatSageEqualToHeaven) [inline Source.lua:172]: This should be `typesVersion`

File: Source.lua [hunk 3]
```diff
@@ -1,229 +1,281 @@
...
-		elseif optype == 1 then --[[ A ]]
-			i.A = bit32.band(bit32.rshift(value, 8), 0xFF)
-		elseif optype == 4 then --[[ AD ]]
-			i.A = bit32.band(bit32.rshift(value, 8), 0xFF)
+
+		local opInfo = opList[opcode + 1]
+		local opName = opInfo[1]
+		local opMode = opInfo[2]
+		local kMode = opInfo[3]
+		local usesAUX = opInfo[4]
```
Comment (TheGreatSageEqualToHeaven) [inline Source.lua:196]: We should just do `usesAux`

File: Source.lua [hunk 4]
```diff
@@ -359,7 +416,7 @@ local function luau_load(module, env)
 
 			while true do
 				local inst = code[pc]
-				local op = inst.opcode
+				local op = inst.code
```
Comment (TheGreatSageEqualToHeaven) [inline Source.lua:419]: We should keep using `opcode` instead of `code`

File: Source.lua [hunk 5]
```diff
@@ -821,8 +874,8 @@ local function luau_load(module, env)
 					local it = stack[inst.A]
 
 					if type(it) ~= "function" then 
-						local loopInstruction = code[pc + inst.D]
-						if generalized_iterators[loopInstruction] == nil then 
+						local loopinstruction = code[pc + inst.D]
```
Comment (TheGreatSageEqualToHeaven) [inline Source.lua:877]: This should still be `loopInstruction`

**Review by TheGreatSageEqualToHeaven — CHANGES_REQUESTED** [review]
Left some requested changes

File: Source.lua (hunk 5 above)
Comment (green-real) [reply Source.lua:877]: oh yeah, didn't mean to do that

**Review by TheGreatSageEqualToHeaven — APPROVED** [review]
LGTM

**Pull request merged**